File: Day12/src/main.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <cmath>
#include <queue>
#include <algorithm>

struct Coord {
    int x = 0;
    int y = 0;
	bool operator==(const Coord& rhs) const { return this->x == rhs.x && this->y == rhs.y;}
	Coord(int x, int y) : x(x), y(y) {}
};

struct GridSquare {
    bool visited = false;
    char c = '\0';
	int dist = 2000000;
	Coord coord{0, 0};
};

class Grid {
public:
    int len_x = 0;
    int len_y = 0;
    GridSquare** squares = nullptr;
    Coord start{0,0};
    Coord end{0,0};
    std::vector<char> path = {};

	void calc_lengths() const {
		// Calculate the path distances starting from the end point
		auto cmp = [](GridSquare* a, GridSquare* b) { return a->dist > b->dist; };
		auto queue = std::priority_queue<GridSquare*, std::vector<GridSquare*>, decltype(cmp)>(cmp);

		int x = this->end.x;
		int y = this->end.y;
		auto curr = this->squares[y * len_x + x];
		curr->dist = 0;
		queue.push(curr);

		// Check in 3x3 area around the point
		// Only continue if its distance is bigger than the distance checked, and is possible
		// Impossible points have distance +infinity
		// Point gets distance of self + 1
		// PriorityQueue check out the lowest points
		std::vector<Coord> coords;

		while (!queue.empty()) {
			GridSquare* current = queue.top();
			current = this->squares[current->coord.y * this->len_x + current->coord.x];
			current->visited = true;
			queue.pop();
			if (current->coord == start) {
				std::cout << "START!!! " << current->dist << std::endl;
			}

			int y_min = current->coord.y == 0 ? current->coord.y : current->coord.y - 1;
			int y_max = current->coord.y == this->len_y - 1 ? current->coord.y : current->coord.y + 1;
			int x_min = current->coord.x == 0 ? current->coord.x : current->coord.x - 1;
			int x_max = current->coord.x == this->len_x - 1 ? current->coord.x : current->coord.x + 1;

			for (int y_inner = y_min; y_inner <= y_max; y_inner++) {
				for (int x_inner = x_min; x_inner <= x_max; x_inner++) {
					if (x_inner == current->coord.x && y_inner == current->coord.y || (pow(x_inner - current->coord.x, 2) + pow(y_inner - current->coord.y, 2)) == 2) continue;
					auto next = this->squares[y_inner * len_x + x_inner];
					if (next->dist <= current->dist) continue;
					if (current->c - next->c != 1 && current->c - next->c != 0 && current->c - next->c != -2) continue;
					if (next->dist > current->dist + 1)
						next->dist = current->dist + 1;
					if (std::find(coords.begin(), coords.end(), next->coord) != coords.end()) {
						continue;
					}
					if (!next->visited) {
						queue.push(next);
						coords.push_back(next->coord);
					}
				}
			}
		}
	}

     ~Grid() {
         delete[] squares;
     }
};

int main(int argc, char** argv) {
	if (argc < 2) {
		std::cerr << "Please provide an argument" << std::endl;
		exit(1);
	}
	std::cout << "Using file: " << argv[1] << std::endl;
    std::ifstream file = std::ifstream(argv[1]);
    std::string line;
    int len_y = 0;
    int len_x = 0;
    Coord start{0, 0};
    Coord end{0, 0};
    Grid grid;

	while (std::getline(file, line)) {
		std::cout << line << std::endl;
		len_x = line.size();
		len_y++;
	}

	file.close();
	int y = 0;
	grid.squares = new GridSquare*[len_y * len_x];

	file = std::ifstream(argv[1]);
    while (std::getline(file, line)) {
        for (int i = 0; i < line.size(); i++) {
            char c = line[i];
            if (c == 'S') {
                start.x = i;
                start.y = y;
                c = 'a';
            } else if (c == 'E') {
                end.x = i;
                end.y = y;
                c = 'z';
            }
            auto* square = new GridSquare(); // oooooh, spoooky heap allocation
            square->c = c;
            square->visited = false;
			square->coord.x = i;
			square->coord.y = y;
            grid.squares[y * len_x + i] = square;
        }
		y++;
    }

    grid.len_x = len_x;
    grid.len_y = len_y;
    grid.start.x = start.x;
    grid.start.y = start.y;
    grid.end.x = end.x;
    grid.end.y = end.y;
    std::cout << "Start: X(" << start.x << "), Y(" << start.y << ")" << std::endl;
    std::cout << "End: X(" << end.x << "), Y(" << end.y << ")" << std::endl;
    //grid.solve(start.x, start.y);
	grid.calc_lengths();
	int shortest_a = 2000000;
	for (int row = 0; row < len_y; row++) {
		for (int col = 0; col < len_x; col++) {
			auto square = grid.squares[row * len_x + col];
			//std::cout << square->dist << ", ";
			if (square->c == 'a' && square->dist < shortest_a) {
				shortest_a = square->dist;
			}
		}
		std::cout << std::endl;
	}
	std::cout << "Shortest A (Part B): " << shortest_a << std::endl;
	file.close();
}

File: Day12/src/main.rs
use std::env;
use std::fs;

#[derive(Clone, Copy)]
struct GridSquare {
    char: char,
    visited: bool,
}

struct Grid {
    len_x: usize,
    squares: Vec<Vec<GridSquare>>,
    start_point: (usize, usize),
    end_point: (usize, usize),
}

impl Grid {
    fn solve(&mut self, path: &mut Vec<char>, x: usize, y: usize) {
        //println!("Solving... p: {}, x: {}, y: {}", path.iter().collect::<String>(), x, y);
        if (x, y) == self.end_point {
            println!(
                "Solved! Path: {}, len = {}",
                path.iter().collect::<String>(),
                path.len()
            );
            return;
        }
        let len = self.squares.len();
        let len_x = self.len_x;
        // x - 1, x + 1, y - 1, y + 1
        let y_min = if y == 0 { y } else { y - 1 };
        let y_max = if y == len - 1 { y } else { y + 1 };
        let x_min = if x == 0 { x } else { x - 1 };
        let x_max = if x == len_x - 1 { x } else { x + 1 };
        
        println!("Y: min({}), max({}), X: min({}), max({})", y_min, y_max, x_min, x_max);
        let curr = self.get_square_mut(x, y);
        curr.visited = true;
        
        for y_inner in y_min..=y_max {
            for x_inner in x_min..=x_max {
                if (x, y) == (x_inner, y_inner) {
                    continue;
                }
                let next = self.get_square_mut(x_inner, y_inner);
                
                if !self.possible(curr, next) {
                    continue;
                }

                //println!("LOOP");
                next.visited = true;
                path.push(next.char);
                self.solve(path, x_inner, y_inner);
                // Backtrack
                next.visited = false;
                path.pop();
            }
        }
    }

    fn get_square(&self, x: usize, y: usize) -> &GridSquare {
        self.squares.get(y).unwrap().get(x).unwrap()
    }

    fn get_square_mut(&mut self, x: usize, y: usize) -> &mut GridSquare {
        &mut self.squares[y][x]
    }

    fn possible(&self, curr: &GridSquare, next: &GridSquare) -> bool {
        !next.visited && (next.char as i16 - curr.char as i16) <= 1
    }

    fn possible_mut(&mut self, curr: &GridSquare, next: &GridSquare) -> bool {
        println!("Possible? Curr: c({}) v({}), Next: c({}) v({})", curr.char, curr.visited, next.char, next.visited);
        !next.visited && (next.char as i16 - curr.char as i16) <= 1
    }
}

fn main() {
    let args: Vec<String> = env::args().collect();
    let file_path = &args[1];
    println!("In file {}", file_path);

    let contents = fs::read_to_string(file_path).expect("Should have been able to read the file");
    let mut len = 0;
    let mut grid_squares: Vec<Vec<GridSquare>> = Vec::new();
    let mut starting_point = (0, 0);
    let mut end_point = (0, 0);
    for y in 0..contents.lines().count() {
        let line = contents.lines().nth(y).unwrap();
        len = line.len();
        if let Some(index) = line.find('S') {
            starting_point = (index, y);
        }

        if let Some(index) = line.find('E') {
            end_point = (index, y);
        }
        let chars: Vec<GridSquare> = line
            .chars()
            .map(|c| GridSquare {
                char: if c != 'S' && c != 'E' {c} else {
                    if c == 'S' {'a'} else {'z'}
                },
                visited: false,
            })
            .collect();
        grid_squares.push(chars);
    }

    let mut grid = Grid {
        len_x: len,
        squares: grid_squares,
        start_point: starting_point,
        end_point,
    };

    let mut path: Vec<char> = Vec::new();
    grid.solve(&mut path, starting_point.0, starting_point.1);

    // Solve with backtracking
    // Start at starting point
    // Iterate over possible options
    // If stuck => Backtrack
    // If found => Add to solved solutions and backtrack
}
